File: frontend/src/app/shared/password-policy.ts
/**
 * The password policy the server enforces on every surface that accepts a new password (register on
 * both sides, reset, set, both self-service changes), mirrored client-side so the screens can say the
 * rule before submit and name the failed rule after. The server is authoritative: it answers
 * `INVALID_REQUEST` for the length rule and {@link PASSWORD_BLOCKED_TERM_CODE} for the blocklist.
 * Pure — no HTTP, no app state — so both core auth services and every password screen share it.
 */

/** The minimum length in characters — length is the primary control, there are no composition rules. */
export const PASSWORD_MIN_LENGTH = 12;
/** The maximum in UTF-8 bytes, bcrypt's input cap — an accented passphrase can exceed it well under 72 characters. */
export const PASSWORD_MAX_BYTES = 72;
/** The service name, blocked in any case on every surface. */
export const PASSWORD_BLOCKED_WORD = 'riviera';
/** An account name shorter than this is not applied as a blocked term — it would match almost anything. */
export const PASSWORD_MIN_ACCOUNT_NAME_LENGTH = 3;
/** The problem `code` the server answers when the password contains a blocked term. */
export const PASSWORD_BLOCKED_TERM_CODE = 'PASSWORD_CONTAINS_BLOCKED_TERM';

/** Shown beside the field before submit, on every screen where a password is chosen. */
export const PASSWORD_POLICY_HINT =
  'At least 12 characters. It can’t contain the name you sign in with or the word “riviera”.';
/** The length rule failed — the server's `INVALID_REQUEST`, or the client-side minimum. */
export const PASSWORD_LENGTH_MESSAGE = 'Choose a password of 12–72 characters.';
/** The byte cap failed client-side, where the character count on screen would contradict the length message. */
export const PASSWORD_TOO_LONG_MESSAGE =
  'That password is too long. Accented letters and emoji each take several of the 72 available characters, so try a shorter one.';
/** The blocklist failed — the server's {@link PASSWORD_BLOCKED_TERM_CODE}, or the client-side check. */
export const PASSWORD_BLOCKED_MESSAGE =
  'That password contains the name you sign in with or the word “riviera”. Choose a different one.';

/** Which rule a candidate password fails, when it fails one. */
export type PasswordPolicyViolation = 'too-short' | 'too-long' | 'blocked';

/** Byte length under UTF-8, which is what the server's 72-byte bcrypt cap actually measures. */
export function passwordByteLength(password: string): number {
  return new TextEncoder().encode(password).length;
}

/** The part of an email before the `@`, lower-cased — the name a tourist signs in with. */
export function emailLocalPart(email: string): string {
  const at = email.indexOf('@');
  return (at < 0 ? email : email.slice(0, at)).toLowerCase();
}

/**
 * The client-side check every password screen runs before spending a request. `accountName` is the
 * name the account signs in with (the email's local part, or the operator username) when the screen
 * knows it; the server always applies it. Rules run in the server's order: length, then the blocklist.
 */
export function passwordPolicyViolation(
  password: string,
  accountName?: string,
): PasswordPolicyViolation | undefined {
  if (password.length < PASSWORD_MIN_LENGTH) {
    return 'too-short';
  }
  if (passwordByteLength(password) > PASSWORD_MAX_BYTES) {
    return 'too-long';
  }
  const lower = password.toLowerCase();
  if (lower.includes(PASSWORD_BLOCKED_WORD)) {
    return 'blocked';
  }
  const name = accountName?.toLowerCase() ?? '';
  if (name.length >= PASSWORD_MIN_ACCOUNT_NAME_LENGTH && lower.includes(name)) {
    return 'blocked';
  }
  return undefined;
}

/** The one message per failed rule, the same wording on every screen. */
export function passwordPolicyMessage(violation: PasswordPolicyViolation): string {
  switch (violation) {
    case 'too-short':
      return PASSWORD_LENGTH_MESSAGE;
    case 'too-long':
      return PASSWORD_TOO_LONG_MESSAGE;
    case 'blocked':
      return PASSWORD_BLOCKED_MESSAGE;
  }
}

File: platform/src/main/java/ai/riviera/platform/PasswordPolicy.java
package ai.riviera.platform;

import ai.riviera.platform.shared.InvalidApiRequestException;

import java.nio.charset.StandardCharsets;
import java.util.Locale;

/**
 * The one password policy every edge surface that accepts a new password enforces — tourist and
 * operator register, reset, set, and both self-service changes — plus the bootstrap credential's
 * length rule. A stateless static helper — no bean, no Spring Security type — so the same rule
 * applies without threading a collaborator through each constructor.
 *
 * <p>Policy (design D-8): {@value #MIN_LENGTH} characters to {@value #MAX_BYTES} bytes (bcrypt's input
 * cap), leading and trailing spaces significant, no composition rules; and the password may not
 * contain, case-insensitively, the service name or the account's own name (the email local part for
 * a tourist, the username for an operator). A length violation throws
 * {@link InvalidApiRequestException} ({@code 400 INVALID_REQUEST}); a blocklist hit throws
 * {@link BlockedPasswordException} ({@code 400 PASSWORD_CONTAINS_BLOCKED_TERM}) so a client can say
 * which rule failed. Length is checked first. Rationale: {@code RESPONSIBILITIES.md} § Platform edge.
 */
final class PasswordPolicy {

	static final int MIN_LENGTH = 12;
	static final int MAX_BYTES = 72;
	/** An account name shorter than this is not applied as a blocked term — it would match almost anything. */
	static final int MIN_ACCOUNT_NAME_LENGTH = 3;
	private static final String SERVICE_NAME = "riviera";

	private PasswordPolicy() {
	}

	/**
	 * Whether a current-password field was supplied at all — the one definition shared by both self-service
	 * change endpoints, so the operator and customer twins cannot drift on what "supplied" means.
	 * The test is <em>empty</em>, never blank: the policy forbids a stored password under
	 * {@value #MIN_LENGTH} characters so {@code ""} can never be a real one, while leading and
	 * trailing spaces are significant and must survive.
	 */
	static boolean isSupplied(String password) {
		return password != null && !password.isEmpty();
	}

	/** Whether {@code password} is within the length bounds — the check the bootstrap credential shares. */
	static boolean hasPermittedLength(String password) {
		int bytes = password.getBytes(StandardCharsets.UTF_8).length;
		return password.length() >= MIN_LENGTH && bytes <= MAX_BYTES;
	}

	/** Enforce the length rule and the service-name block before any encode/write. */
	static void validate(String password) {
		if (!hasPermittedLength(password)) {
			throw new InvalidApiRequestException("password outside the permitted length");
		}
		if (contains(password, SERVICE_NAME)) {
			throw new BlockedPasswordException();
		}
	}

	/**
	 * {@link #validate(String)} plus the account's own name as a blocked term: the email local part for
	 * a tourist, the username for an operator. A name under {@value #MIN_ACCOUNT_NAME_LENGTH} characters is skipped.
	 */
	static void validate(String password, String accountName) {
		validate(password);
		if (accountName.length() >= MIN_ACCOUNT_NAME_LENGTH && contains(password, accountName)) {
			throw new BlockedPasswordException();
		}
	}

	/** The part of {@code email} before the {@code @}, lower-cased — the tourist's account name. */
	static String emailLocalPart(String email) {
		int at = email.indexOf('@');
		String localPart = at < 0 ? email : email.substring(0, at);
		return localPart.toLowerCase(Locale.ROOT);
	}

	private static boolean contains(String password, String term) {
		return password.toLowerCase(Locale.ROOT).contains(term.toLowerCase(Locale.ROOT));
	}

}
